src/querybuilder/components/MetricParamEditor.tsx

```diff
@@ -0,0 +1,52 @@
+import React, { useState } from "react";
+
+import { SelectableValue, toOption } from "@grafana/data";
+import { Select } from "@grafana/ui";
+
+import { PrometheusDatasource } from "../../datasource";
+import { getOperationParamId } from "../shared/operationUtils";
+import { QueryBuilderOperationParamEditorProps } from "../shared/types";
+import { PromVisualQuery } from "../types";
+
+import { getMetrics } from "./PromQueryBuilder";
+
+export function MetricParamEditor({
+  onChange, index, operationIndex, value, query, datasource
+}: QueryBuilderOperationParamEditorProps) {
+  const [state, setState] = useState<{
+    options?: Array<SelectableValue<any>>;
+    isLoading?: boolean;
+  }>({});
+
+  return (
+    <Select
+      inputId={getOperationParamId(operationIndex, index)}
+      autoFocus={value === "" ? true : undefined}
+      openMenuOnFocus
+      onOpenMenu={async () => {
+        setState({ isLoading: true });
+        const options = await loadGroupByLabels(query, datasource as PrometheusDatasource);
+        setState({ options, isLoading: undefined });
+      }}
+      isLoading={state.isLoading}
+      allowCustomValue
+      noOptionsMessage="No labels found"
+      loadingMessage="Loading labels"
+      options={state.options}
+      value={toOption(value as string)}
+      onChange={(value) => onChange(index, value.value!)}
```

Comment:
it would be better to move all logic to consts and functions to better readable code